### ISSUE TITLE
ORDER BY WITH FILL + DateTime64

### ISSUE THREAD

**Issue by Slach** [ClickHouse/ClickHouse#12405]
**Describe the bug**
Exception when try to use `toDateTime64` and `ORDER BY ... WITH FILL`
**How to reproduce**
```sql
 SELECT n, source FROM (    
   SELECT toDateTime64(number*1000, 3) AS n, 'original' AS source    
   FROM numbers(10) WHERE number % 3 = 1 
) 
ORDER BY n WITH FILL STEP 1000;
```

* Which ClickHouse server version to use
20.5.2.7

**Expected behavior**
fill with step 1000 nanoseconds?

**Error message and/or stacktrace**
```
Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: Add different decimal fields.
```

**Comment by Slach** [comment]
@CurtizJ could you look to this issue?

**Comment by alexey-milovidov** [comment]
CC @Enmk

**Comment by alexey-milovidov** [comment]
It also does not work when I try to convert types:

```
milovidov-desktop :) SELECT n, source FROM (　   SELECT toDateTime64(number*1000, 3) AS n, 'original' AS source　   FROM numbers(10) WHERE number % 3 = 1　)　ORDER BY n WITH FILL STEP toDateTime64(1000, 3);

SELECT
    n,
    source
FROM 
(
    SELECT
        toDateTime64(number * 1000, 3) AS n,
        'original' AS source
    FROM numbers(10)
    WHERE (number % 3) = 1
)
ORDER BY n ASC WITH FILL STEP toDateTime64(1000, 3)

Query id: fcab6c79-58af-4d53-8065-818a5b37a0b1


0 rows in set. Elapsed: 0.012 sec. 

Received exception from server (version 21.3.1):
Code: 475. DB::Exception: Received from localhost:9000. DB::Exception: Incompatible types of WITH FILL expression values with column type DateTime64(3).
```

**Comment by CurtizJ** [comment]
Implemented in #24016.

**Comment by Slach** [comment]
@MaxWk thanks  a lot for your PR
